ArduCopter/GCS_Mavlink.cpp

```diff
@@ -1054,11 +1054,24 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
             break;
         }
 
+        bool roll_rate_ignore   = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_ROLL_RATE_IGNORE;
+        bool pitch_rate_ignore  = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_PITCH_RATE_IGNORE;
+        bool yaw_rate_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_BODY_YAW_RATE_IGNORE;
+        bool throttle_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_THROTTLE_IGNORE;
+        bool attitude_ignore    = packet.type_mask & ATTITUDE_TARGET_TYPEMASK_ATTITUDE_IGNORE;
+
         // ensure type_mask specifies to use attitude and thrust
-        if ((packet.type_mask & ((1<<7)|(1<<6))) != 0) {
+        if (throttle_ignore) {
             break;
         }
 
+        Quaternion attitude_quat;
+        if (attitude_ignore) {
+            attitude_quat = Quaternion(0.0, 0.0, 0.0, 0.0);
+        } else {
+            attitude_quat = Quaternion(packet.q[0],packet.q[1],packet.q[2],packet.q[3]);
+        }
```

Comment:
Yes, I agree. There are a number of issues here to discuss.

I used the 0,0,0,0 because that was something that has been defined in Mavlink here:
https://mavlink.io/en/messages/common.html#ATTITUDE_QUATERNION
"send [0, 0, 0, 0] if field not supported)"

Reply:
> Passing the Quat(0,0,0,0) to switch off of in modeguided makes this seem like a mistake at first. I'd say just pass attitude_ignore into guided??

Yes, this is another question to answer. This approach would obviously need a comment before we leave it like this. But passing a boolean into guided is a valid option, feels a bit ugly to me. We start to get to the point where we may as well pass the whole bitmask into guided.

Reply:
> > Passing the Quat(0,0,0,0) to switch off of in modeguided makes this seem like a mistake at first. I'd say just pass attitude_ignore into guided??
> 
> Yes, this is another question to answer. This approach would obviously need a comment before we leave it like this. But passing a boolean into guided is a valid option, feels a bit ugly to me. We start to get to the point where we may as well pass the whole bitmask into guided.

Yeah position_target_int is like that with passing so many variables. Possibly, the solution is to create a structure that holds all the needed variables for these functions and pass those into guided. Then we wouldn't need to change the function definitions which every change or additional feature.

Reply:
I would like to leave it like this and get it in and work on cleaning things up after.

Reply:
That's fine lets leave this open till it is handled by a later PR

Reply:
We still need to check for this to ensure we can normalise the quaternion safely.

Reply:
It is also worth noting that we often use zero to denote a parameter is not defined.

Reply:
I think I handled this with the check for unit length input